motion/hybridpath_controller/hybridpath_controller/adaptive_backstep.py

```diff
@@ -0,0 +1,86 @@
+import numpy as np
+
+class AdaptiveBackstep:
+    def __init__(self, kappa = 0.5):
+        self.init_system(kappa)
+
+    def init_system(self, kappa):
+
+        I = np.eye(3)
+
+        K_1 = np.diag([10, 10, 10])
+        self.K_1_tilde = K_1 + kappa*I
+        self.K_2 = np.diag([60, 60, 30])
+        self.tau_max = np.array([41.0, 50.0, 55.0]) # Må tilpasses thrusterne
+
+        ## Forenklet modell ## Bør også endres
+        m = 50
+        self.M = np.diag([m, m, m])
+        self.D = np.diag([10, 10, 5])
+
+    def control_law(self, eta, nu, w, v_ref, dt_v_ref, dtheta_v_ref, eta_d, dtheta_eta_d, ddtheta_eta_d): # dtheta == ds
+        _, R_trps = self.R(eta[2])
+        S = self.S(nu[2])
+
+        eta_error = eta - eta_d
+        eta_error[2] = self.ssa(eta_error[2])
+
+        z1 = R_trps @ eta_error
+        alpha1 = -self.K_1_tilde @ z1 + R_trps @ dtheta_eta_d * v_ref
+
+        z2 = nu - alpha1
+
+        sigma1 = self.K_1_tilde @ (S @ z1) - self.K_1_tilde @ nu - S @ (R_trps @ dtheta_eta_d) * v_ref + R_trps @ dtheta_eta_d * dt_v_ref
+
+        dtheta_alpha1 = self.K_1_tilde @ (R_trps @ dtheta_eta_d) + R_trps @ ddtheta_eta_d * v_ref + R_trps @ dtheta_eta_d * dtheta_v_ref
+
+        # Control law ## Må endres når system-matrisene endres
+        tau = -self.K_2 @ z2 + self.D @ nu + self.M @ sigma1 + self.M @ dtheta_alpha1 * (v_ref + w)
+
+        # Add constraints to tau #
+
+        if np.absolute(tau[0]) > self.tau_max[0] or np.absolute(tau[1]) > self.tau_max[1] or np.absolute(tau[2]) > self.tau_max[2]:
+            if np.absolute(tau[0]) > self.tau_max[0]:
+                tau[2] = np.sign(tau[2]) * np.absolute(self.tau_max[0] / tau[0]) * np.absolute(tau[2])
+                tau[1] = np.sign(tau[1]) * np.absolute(self.tau_max[0] / tau[0]) * np.absolute(tau[1])
+                tau[0] = np.sign(tau[0]) * self.tau_max[0]
+            if np.absolute(tau[1]) > self.tau_max[1]:
+                tau[2] = np.sign(tau[2]) * np.absolute(self.tau_max[1] / tau[1]) * np.absolute(tau[2])
+                tau[0] = np.sign(tau[0]) * np.absolute(self.tau_max[1] / tau[1]) * np.absolute(tau[0])
+                tau[1] = np.sign(tau[1]) * self.tau_max[1]
+            if np.absolute(tau[2]) > self.tau_max[2]:
+                tau[1] = np.sign(tau[1]) * np.absolute(self.tau_max[2] / tau[2]) * np.absolute(tau[1])
+                tau[0] = np.sign(tau[0]) * np.absolute(self.tau_max[2] / tau[2]) * np.absolute(tau[0])
+                tau[2] = np.sign(tau[2]) * self.tau_max[2]
+        return tau
```

Comment:
It's not immediately clear to me how, but it smells like this could be refactored into a more manageable method.

You also don't need the nested ifs since these all perform redundant checking